packages/unified-selection/src/unified-selection/HiliteSetProvider.ts

```diff
@@ -54,49 +56,67 @@ class HiliteSetProviderImpl implements HiliteSetProvider {
   private _queryExecutor: IECSqlQueryExecutor;
   private _metadataProvider: IMetadataProvider;
   // Map between a class name and its type
-  private _classRelationCache: Map<string, InstanceIdType>;
+  private _classRelationCache: Map<string, InstanceIdType | Promise<InstanceIdType>>;
 
   constructor(props: HiliteSetProviderProps) {
     this._queryExecutor = props.queryExecutor;
     this._metadataProvider = props.metadataProvider;
-    this._classRelationCache = new Map<string, InstanceIdType>();
+    this._classRelationCache = new Map();
   }
 
   /**
    * Get hilite set iterator for supplied `Selectables`.
    */
-  public async *getHiliteSet({ selectables }: { selectables: Selectables }): AsyncIterableIterator<HiliteSet> {
-    // Map between element ID types and EC instance IDs
-    const keysByType = new Map<InstanceIdType, string[]>();
-
-    for (const entry of selectables.instanceKeys) {
-      for (const key of entry[1]) {
-        await this.addKeyByType({ className: entry[0], id: key }, keysByType);
-      }
-    }
-
-    for (const entry of selectables.custom) {
-      for await (const key of entry[1].loadInstanceKeys()) {
-        await this.addKeyByType(key, keysByType);
-      }
-    }
+  public getHiliteSet(props: { selectables: Selectables }): AsyncIterableIterator<HiliteSet> {
+    const obs = this.getHiliteSetObservable(props);
+    return eachValueFrom(obs);
+  }
 
-    yield {
-      models: await this.getHilitedModels(keysByType),
-      subCategories: await this.getHilitedSubCategories(keysByType),
-      elements: await this.getHilitedElements(keysByType),
+  /**
+   * Returns a "hot" observable of hilite sets.
+   */
+  private getHiliteSetObservable({ selectables }: { selectables: Selectables }): Subject<HiliteSet> {
+    const instancesByType = this.getInstancesByType(selectables);
+    const observables: { [key in keyof HiliteSet]?: Observable<string> } = {
+      models: this.getHilitedModels(instancesByType),
+      subCategories: this.getHilitedSubCategories(instancesByType),
+      elements: this.getHilitedElements(instancesByType),
     };
-  }
 
-  private async addKeyByType(key: SelectableInstanceKey, map: Map<string, string[]>) {
-    const keyType = await this.getType(key);
-    let array = map.get(keyType);
-    if (!array) {
-      array = [];
-      map.set(keyType, array);
-    }
+    let hiliteSet: HiliteSet = { models: [], subCategories: [], elements: [] };
+    let lastEmitTime = performance.now();
+    const subject = new Subject<HiliteSet>();
+    const subscriptions = (["models", "subCategories", "elements"] as const).map((key) =>
+      observables[key]!.subscribe({
+        next(val) {
+          hiliteSet[key].push(val);
+          if (performance.now() - lastEmitTime < 20) {
```

Comment:
let's define a `const` for this value

Reply:
For `observables[key]!` or the result of `subscribe`?

Reply:
`20`

Reply:
Oh, that's right. I guess, it's Monday for me as well today :D
Extracted to constant